src/Workspaces/CSharp/Portable/Classification/Worker.cs

```diff
@@ -262,5 +265,52 @@ private void ClassifyDisabledText(SyntaxTrivia trivia, SyntaxTriviaList triviaLi
                 AddClassification(trivia, ClassificationTypeNames.ExcludedCode);
             }
         }
+
+        private void ClassifyLiteralAndSuffix(TextSpan tokenSpan, string literalClassification, int suffixLength)
+        {
+            var split = tokenSpan.End - suffixLength;
+            AddClassification(TextSpan.FromBounds(tokenSpan.Start, split), literalClassification);
+            AddClassification(TextSpan.FromBounds(split, tokenSpan.End), ClassificationTypeNames.Keyword);
+        }
+
+        private static bool TryGetNumericLiteralSuffix(string tokenText, out int suffixLength)
+        {
+            // Walk left for as long as we encounter valid suffixes (both integer and real) and count the number of characters on the way.
+            // The lexer ensures that the token only ends in valid suffixes; invalid suffix characters are part of a different token.
+
+            // § 6.4.5.3: '0x' and '0b' mean it must be an integer literal, so we can disregard the real suffixes.
+            var allowRealTypeSuffixes = !IsHexOrBinaryInteger(tokenText);
+
+            suffixLength = 0;
+            for (var i = tokenText.Length - 1; i >= 0 && IsValidSuffixCharacter(tokenText[i], allowRealTypeSuffixes); i--)
+            {
+                suffixLength++;
+            }
+
+            return suffixLength > 0;
+
+            static bool IsHexOrBinaryInteger(string tokenText)
+            {
+                if (tokenText.Length < 2)
+                {
+                    return false;
+                }
+
+                return tokenText[0] is '0' && ToUppercase(tokenText[1]) is 'X' or 'B';
+            }
+
+            static bool IsValidSuffixCharacter(char c, bool allowRealTypeSuffixes)
+            {
+                return ToUppercase(c) switch
```

Comment:
1. honestly, no need for ToUpperCase.  Can just list the cases.
2. i'm genuinly fine with just checking for the 5 possible suffixes (10 including different cases).  The need to check if it's real or not just doesn't matter in practice.

Reply:
Re 2: The real-ness is not the interesting thing here, it's the conflict between `F` as a hex digit (only valid in integer literals, but then it can't be a suffix) and `F` as a suffix (only valid in real literals, but those cannot be hex). If we don't differentiate between these, `0x8fdul` is classified as number `0x8` and suffix `fdul`. So we check whether the token is definitely an integer literal, in which case D/F/M simply cannot be suffixes. I renamed one of the variables to hopefully make this a bit clearer.